### ISSUE TITLE
MSB4011 is not suppressable by MSBuildWarningsAsMessages

### ISSUE THREAD

**Issue by rainersigwald** [Microsoft/msbuild#4038]
Originally reported at https://developercommunity.visualstudio.com/content/problem/400893/using-msbuildwarningsasmessages-in-directorybuildp.html.

> I have added Directory.Build.props to the same folder as my solution with content:
> 
> ```xml
> <Project>
> <PropertyGroup> 
> <MSBuildWarningsAsMessages>MSB4011</MSBuildWarningsAsMessages>
> </PropertyGroup>
> </Project>
> ```
> 
> The problem is that I still get the MSB4011 warnings when I compile.
> 
> I know that file is found because if I change the <PropertyGroup> tag to <PropertyGroup1>, I get an error when compiling.
> 
> Why isn't it working?
> 
> I am using visual studio professional 2017, 15.9.3. The projects are C++.

**Comment by rainersigwald** [comment]
I think this is happening for reasons related to #3295: the warning path for evaluation-related warnings is distinct from normal execution-time warnings.

In this case, the final value of `MSBuildWarningsAsMessages` can't be known mid-evaluation, so I'm not sure what could be done to fix this. We could maybe inspect the current value just before logging a warning, but that seems heavy-handed.

**Comment by rainersigwald** [comment]
Smallest repro I could think of: 
[repro.zip](https://github.com/Microsoft/msbuild/files/2722067/repro.zip)

**Comment by jeffkl** [comment]
Some warnings are logged by the same code that parse the codes to suppress in your `Directory.Build.props`.  To suppress this warning (not recommended if you can just fix the double import) you can use the `/nowarn` command-line argument for builds run via `msbuild.exe`.  But there is no way to suppress it in Visual Studio.

**Comment by botrif** [comment]
It's impossible to avoid this warning when using new SDK-style projects if T4 templates are to be transformed because the template targets must be included _after_ the CSharp targets, and with SDK-style projects the CSharp targets get automatically included at the end of the csproj. I'd love to know of any workaround for this double-include, or at least how to get rid of this warning in the IDE

**Comment by rainersigwald** [comment]
@botrif I'm not sure I understand the whole situation, but can you expand the Sdk imports into their explicit form, like:

```xml
<Project>
    <Import Project="Sdk.props" Sdk="Microsoft.NET.Sdk" />

    <PropertyGroup>
        <TargetFramework>net46</TargetFramework>
    </PropertyGroup>

    <Import Project="Sdk.targets" Sdk="Microsoft.NET.Sdk" />

    <Import Project="T4 stuff" />
</Project>
```

?

**Comment by alexdrl** [comment]
We are facing the same problem, but we are trying to make this warnings as error to stop this from happening earlier.

**Comment by bhupeshpant19jan** [comment]
We are also hitting the same issue. This is really discouraging.

**Comment by chucklu** [comment]
I encounter the MSB4011 warning, and it blocked the build process of Visual Studio.
After review the history of .csproj file, and find someone else add the following section in it.
```xml
 <NuGetPackageImportStamp>209d4829</NuGetPackageImportStamp>
```
I tried to remove the section, and rebuild, it works.

**Comment by lapinott** [comment]
Same issue here.

Any chance suppressing this warning (or any MSB... for that matter) will be featured?

Full issue here (self-explanatory) : https://docs.microsoft.com/en-us/answers/questions/823721/visual-studio-34msb401134-build-warning-i-want-to.html

Would something similar to "#pragma once" reasonable for property sheets?

**Comment by michael-hawker** [comment]
It seems odd that this is special cased to not be suppressible? We're also using our props for dependency management across projects, those projects reference each other, so I don't think we can work around having them included multiple times.

So these warnings are just noise masking more important warnings in VS.

**Comment by rainersigwald** [comment]
> If one wants to override a default target of the standard build process it is necessary to explicitly add `<Import Project="Sdk.targets" Sdk="Microsoft.NET.Sdk" />` and after that define what is to be overridden as outlined [here](https://github.com/MicrosoftDocs/visualstudio-docs/issues/3902). However, this line results in a `MSB4011` warning.

That should not be true--did you add that explicit import without removing the `Sdk="Microsoft.NET.Sdk` from the top-level `<Project>` element?

The docs on this weren't super clear so I submitted MicrosoftDocs/visualstudio-docs#8444.

**Comment by ackh** [comment]
@rainersigwald You are right, I simply overlooked that I need to remove `Sdk="Microsoft.NET.Sdk"` from the `Project` node. Doing so prevents my problem entirely. Thanks for your quick reaction to my comment!
